src/afa/afa.cc

```diff
@@ -1003,41 +1002,127 @@ void Mata::Afa::construct(
+    bool remove_state_map = false;
+    if (nullptr == state_map) {
+        state_map = new StringToStateMap();
+        remove_state_map = true;
+    }
 
-  Mata::Nfa::OnTheFlyAlphabet alphabet(*symbol_map);
+    // a lambda for translating state names to identifiers
+    auto get_state_name = [&state_map, &aut](const std::string& str) {
+        if (!state_map->count(str)) {
+            State state = aut.add_new_state();
+            state_map->insert({str, state});
+            return state;
+        } else {
+            return (*state_map)[str];
+        }
+    };
+
+    // a lambda for cleanup
+    auto clean_up = [&]() {
+        if (remove_state_map) { delete state_map; }
+    };
+
+    // lambda returning true if node is operator of given type
+    auto is_node_operator =
+            [](const FormulaNode& node, FormulaNode::OperatorType type) -> bool {
+        return node.is_operator() && node.operator_type == type;
+    };
+
+    // lambda creates Node from set of strings which are state names
+    auto create_node = [get_state_name](
+            const std::unordered_set<std::string>& states_names) -> Node {
+        Node tgt_node;
+        for (const std::string& s : states_names)
+        {
+            tgt_node.insert(get_state_name(s));
+        }
+
+        return tgt_node;
+    };
+
+    for (const auto& str : inter_aut.initial_formula.collect_node_names())
+    {
+        State state = get_state_name(str);
+        aut.initialstates.insert(state);
+    }
 
-	try
-	{
-		construct(aut, parsec, &alphabet, state_map);
-	}
-	catch (std::exception&)
-	{
-		release_res();
-		throw;
-	}
+    for (const auto& str : inter_aut.final_formula.collect_node_names())
+    {
+        State state = get_state_name(str);
+        aut.finalstates.insert(state);
+    }
 
-	release_res();
-} // construct(StringToSymbolMap) }}}
+    for (const auto& trans : inter_aut.transitions)
+    {
+        State src_state = get_state_name(trans.first.name);
+        if (trans.second.node.is_operand() && trans.second.node.operand_type == FormulaNode::SYMBOL)
+        {
+            Symbol symbol = alphabet->translate_symb(trans.second.node.name);
+            aut.add_trans(src_state, symbol, Node());
+            continue;
+        }
+        else if (trans.second.children.size() != 2)
+        {
+            clean_up();
+            if (trans.second.children.size() == 1)
+            {
+                throw std::runtime_error("Epsilon transitions not supported");
+            }
+            else
+            {
+                throw std::runtime_error("Invalid transition");
+            }
+        }
+
+        assert(is_node_operator(trans.second.node, FormulaNode::AND));
+        assert(trans.second.children.front().node.is_operand());
+        Symbol symbol = alphabet->translate_symb(trans.second.children.front().node.name);
+
+        const FormulaGraph* act_graph = &trans.second.children[1];
```

Comment:
Is there a better name? Don't know what `act` is. Actual? Maybe use `curr_` then, it is more established IMO.

Reply:
done